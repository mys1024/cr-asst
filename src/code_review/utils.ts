import type { CompletionUsage, CompletionStats } from './ai/completion';

export function usageToString(usage?: CompletionUsage) {
  return [
    '[usage]',
    [
      `promptTokens: ${usage?.promptTokens ? usage.promptTokens : 'N/A'}`,
      `completionTokens: ${usage?.completionTokens ? usage.completionTokens : 'N/A'}`,
      `totalTokens: ${usage?.totalTokens ? usage.totalTokens : 'N/A'}`,
    ].join(', '),
  ].join(' ');
}

export function statsToString(stats: CompletionStats) {
  return [
    '[stats]',
    [
      `timeToFirstToken: ${(stats.timeToFirstToken / 1000).toFixed(2)}s`,
      `timeToFinish: ${(stats.timeToFinish / 1000).toFixed(2)}s`,
      `tokensPerSecond: ${stats.tokensPerSecond ? `${stats.tokensPerSecond.toFixed(2)} tokens/s` : 'N/A'}`,
    ].join(', '),
  ].join(' ');
}
