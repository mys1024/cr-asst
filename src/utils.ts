import { execa } from 'execa';

export async function gitDiff(options: {
  src: string;
  dst: string;
  excludePaths: string[];
}): Promise<string> {
  const { src, dst, excludePaths } = options;

  const { stdout } = await execa('git', [
    'diff',
    src,
    dst,
    '--',
    '.',
    ...excludePaths.map((path) => `:!${path}`),
  ]);

  return stdout;
}
