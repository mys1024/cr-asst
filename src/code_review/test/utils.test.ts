import { describe, expect, it } from 'vitest';
import { usageToString, statsToString } from '../utils';

describe('usageToString()', () => {
  it('case 1', async () => {
    expect(
      usageToString({
        inputTokens: 321,
        cachedInputTokens: 123,
        outputTokens: 3245,
        reasoningTokens: 2345,
        totalTokens: 7000,
      }),
    ).toMatchSnapshot();
  });
});

describe('statsToString()', () => {
  it('case 1', async () => {
    const startedAt = 1741661141870;
    const firstTokenReceivedAt = 1741661156970;
    const finishedAt = 1741661185963;
    expect(
      statsToString({
        startedAt,
        firstTokenReceivedAt,
        finishedAt,
        timeToFirstToken: firstTokenReceivedAt - startedAt,
        timeToFinish: finishedAt - startedAt,
        tokensPerSecond: 6554 / ((finishedAt - startedAt) / 1000),
      }),
    ).toMatchSnapshot();
  });
});
