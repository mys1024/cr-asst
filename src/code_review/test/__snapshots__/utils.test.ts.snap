// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`statsToString() > case 1 1`] = `"[STATS] timeToFirstToken: 15.10s, timeToFinish: 44.09s, tokensPerSecond: 148.64 tokens/s"`;

exports[`usageToString() > case 1 1`] = `"[USAGE] inputTokens: 321, cachedInputTokens: 123, uncachedInputTokens: 198, outputTokens: 3245, reasoningTokens: 2345, totalTokens: 7000"`;
